### PR TITLE
Enable DNSLink by default

### LINKED ISSUE #288: experimental DNSLINK support doesn't work with /ipns/ path
I have a DNS TXT entry for ggr.com:

````
~ ❯❯❯ dig ggr.com TXT                                                       Mon 2017-10-09 10:47:56 -0700

; <<>> DiG 9.4.2-P2 <<>> ggr.com TXT
;; global options:  printcmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 6017
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;ggr.com.                       IN      TXT

;; ANSWER SECTION:
ggr.com.                60      IN      TXT     "ALIAS for home.ggr.com"
ggr.com.                3600    IN      TXT     "dnslink=/ipns/QmVTofH3UuafnKihbGa6whtsLaH7YyqMWiR9Pzha9xQYuK"

;; Query time: 90 msec
;; SERVER: 127.0.0.1#53(127.0.0.1)
;; WHEN: Mon Oct  9 10:47:59 2017
;; MSG SIZE  rcvd: 133

~ ❯❯❯                                                                       Mon 2017-10-09 10:47:59 -0700
````

Yet, when I visit https://ggr.com using Firefox 56.0 on OpenBSD 6.1-stable-amd64 with ipfs-companion 0.4.11-dev I am NOT redirected to the IPFS version of the site as I had hoped/expected I would be.

I have HSTS enabled on ggr.com. I have more than one TXT DNS entry. I wonder if one of those is causing problems. In any case, even though its an experimental feature, I figured a 100% repro example problem case might be useful.

### PR DESCRIPTION
**Pull request by lidel**
> ## Latest update: https://github.com/ipfs-shipyard/ipfs-companion/pull/558#issuecomment-417081920

Let's make web more robust by making DNSLink ready for mainstream use!

This PR enables efficient DNSLink support by implementing additional lookup strategy from #548.
It also makes it the default for new installs of IPFS Companion.

#### Iteration 2

Applied changes from https://github.com/ipfs-shipyard/ipfs-companion/pull/558#issuecomment-417081920:

> ### TL;DR 
> - Website Owner
>    - publish [DNS TXT record](https://docs.ipfs.io/guides/concepts/dnslink/) and make sure `X-Ipfs-Path` HTTP header is returned along with resources. The header is already returned by HTTP Gateway provided by go-ipfs, just make sure it is not filtered out.
> - Companion Extension
>   - execute DNSLink async lookup  on cache miss and sync if `X-Ipfs-Path` found in  `onHeadersReceived` of initial main request
> 
>  
screenshot_18

### The `dnslink` (boolean flag) is replaced by `dnslinkPolicy`  (select list)

My current plan is to simplify language around this and rename three states for DNSLink Support:
- **disabled**
- **best-effort**  – check cache or (this is new) async lookup in the background  to populate cache
- **enabled** – check cache or do sync (blocking) lookup before sending any HTTP request

Note: if "Detect X-Ipfs-Path header" is enabled (and it will be enabled by default) it will trigger blocking lookup even if **best-effort** is picked.

To ensure DNSLink adoption all old users will be migrated from `dnslink` to `dnslinkPolicy=best-effort` (it can be later changed it in _Preferences_)

See [/docs/dnslink.md](https://github.com/ipfs-shipyard/ipfs-companion/blob/ab72d12f84c0c12b0ecc32e8c256fcad7b38b5ce/docs/dnslink.md#dnslink-support-in-ipfs-companion) for "Read more" explanation of each policy.

### DNS TXT Lookup Result Cache

Switched to in-memory [lru-cache](https://www.npmjs.com/package/lru-cache) with size 1000 and max-age 12h. 

### Detect `X-Ipfs-Path`  (HTTP Header)

Detect `X-Ipfs-Path` and redirect to a valid `/ipfs/` path 
(or `/ipns/` if `dnslinkPolicy=best-effort` is enabled as well).

See [/docs/x-ipfs-path.md](https://github.com/ipfs-shipyard/ipfs-companion/blob/ab72d12f84c0c12b0ecc32e8c256fcad7b38b5ce/docs/x-ipfs-path-header.md#x-ipfs-path-header-support-in-ipfs-companion) for why this is useful. 

### Recover from HTTP Errors

Let's talk about the worst case scenario for `dnslinkPolicy=best-effort`

-  Visiting a new website 
   - DNSLink for FQDN is not in cache
   - remote HTTP server is offline, making it impossible to check if `X-Ipfs-Path` is present

Companion will detect connection error via [`onErrorOccurred`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onErrorOccurred) and check if website can be loaded from IPFS (forced DNSLink lookup using FQDN from HTTP URL).

If DNSLink is present in DNS it is cached (to make all following requests instant by removing HTTP transport for FQDN) and website will be opened from IPFS in a new tab (We can't redirect from `onErrorOccurred`).

Demo: recovering from HTTP connection timeout (httpd is offline):
>  screenshot_11

### Other

- `dnslink.js` cleanup
  - better names
  - added DNSLink lookup fallback to publick gateway if API is offline or running with js-ipfs as embedded gateway
- Added a lot of tests for DNSLink and split `ipfs-request.test.js` as it grew too much.
- Recover on HTTP connection error
- closes https://github.com/ipfs-shipyard/ipfs-companion/issues/548
- closes #288

## Known Issues

- <del>go-ipfs fails to find DNSLink for `ipfs.io` due to `could not resolve name (recursion limit exceeded)` </del>  Solved in https://github.com/ipfs-shipyard/ipfs-companion/pull/558/commits/4be35040d7078c908b20f014c9eb9975e74a1153
  - upstream bug: https://github.com/ipfs/go-ipfs/issues/4293
- `/api/v0/dns does not return ttl

#### Iteration 1

<details>
  <summary>[Collapsed outdated notes] </summary>

> ### TL;DR 
> - Website Owner
>    - publish [DNS TXT record](https://docs.ipfs.io/guides/concepts/dnslink/) and make sure `X-Ipfs-Path` HTTP header is returned along with resources. The header is already returned by HTTP Gateway provided by go-ipfs, just make sure it is not filtered out.
> - Companion Extension
>   - execute DNSLink lookup  on cache miss if `X-Ipfs-Path` found in  `onHeadersReceived` of initial main request
> 
>  screenshot_10

### The `dnslink` (boolean flag) is replaced by `dnslinkPolicy`  (select list)
   It accepts three states described in [/docs/dnslink.md](https://github.com/ipfs-shipyard/ipfs-companion/blob/44f6854aa379a6d3d684138b009552994b1dd142/docs/dnslink.md):
  - `false` – same role as old  `dnslink=false`  (no DNSLink redirects)
  - `detectIpfsPathHeader` – **the new default**, performs DNS TXT lookup only if  `X-Ipfs-Path` is found in  `onHeadersReceived` of initial request
  - `eagerDnsTxtLookup` - `true` – same role as old `dnslink=true`  (DNS TXT lookup for every new hostname)

### DNS TXT Lookup Result Cache

Switched to in-memory [lru-cache](https://www.npmjs.com/package/lru-cache) with size 1000 and max-age 12h. 

### Detect `X-Ipfs-Path`  (HTTP Header)

Detect `X-Ipfs-Path` and redirect to a valid `/ipfs/` path 
(or `/ipns/` if `dnslinkPolicy=detectIpfsPathHeader` is enabled as well).

See [/docs/x-ipfs-path.md](https://github.com/ipfs-shipyard/ipfs-companion/blob/44f6854aa379a6d3d684138b009552994b1dd142/docs/x-ipfs-path-header.md) for why this is useful. 

### Recover from HTTP Errors

Let's talk about the worst case scenario for `dnslinkPolicy=detectIpfsPathHeader`

-  Visiting a new website 
   - DNSLink for FQDN is not in cache
   - remote HTTP server is offline, making it impossible to check if `X-Ipfs-Path` is present

Companion will detect connection error via [`onErrorOccurred`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onErrorOccurred) and check if website can be loaded from IPFS (forced DNSLink lookup using FQDN from HTTP URL).

If DNSLink is present in DNS it is cached (to make all following requests instant by removing HTTP transport for FQDN) and website will be opened from IPFS in a new tab (We can't redirect from `onErrorOccurred`).

Demo: recovering from HTTP connection timeout (httpd is offline):
>  screenshot_11

### Other

- `dnslink.js` cleanup
  - better names
  - added DNSLink lookup fallback to publick gateway if API is offline or running with js-ipfs as embedded gateway
- Added a lot of tests for DNSLink and split `ipfs-request.test.js` as it grew too much.
- Recover on HTTP connection error
- closes https://github.com/ipfs-shipyard/ipfs-companion/issues/548
- closes #288

## Known Issues

- <del>go-ipfs fails to find DNSLink for `ipfs.io` due to `could not resolve name (recursion limit exceeded)` </del>  Solved in https://github.com/ipfs-shipyard/ipfs-companion/pull/558/commits/4be35040d7078c908b20f014c9eb9975e74a1153
  - upstream bug: https://github.com/ipfs/go-ipfs/issues/4293

</details>

----

cc @alanshaw @olizilla  @lgierth @diasdavid  @kyledrake – any thoughts before merging?
(I want to merge & ship to beta channel for initial feedback and community testing)

### CODE REVIEW

**Comment by lidel** [comment]
I gathered some initial feedback and I feel this needs more UX polish:
- DNSLink policy names are too vague
- Explaining policies  raises even more questions
- Mixing X-Ipfs-Path with DNSLink does not help

My current plan is to simplify language around this and rename three states for DNSLink Support:
- **Disabled**
- **Lazy** (is "Opportunistic|" a better word?)  – check cache or DO async lookup in the background 
- **Enabled** – check cache or do sync (blocking) lookup before sending any HTTP request

If "Detect X-Ipfs-Path header" is enabled (and it will be enabled by default) it will trigger blocking lookup even if **Lazy** is picked.

**Comment by lidel** [comment]
Update: simplified options:



"best-effort" executes `preloadDnslink(url)` for every request in async+throttled manner, so it is a safe default even for users running with remote APIs over HTTP:

https://github.com/ipfs-shipyard/ipfs-companion/blob/21913bfe1c7f757402c338bd518c5083e0146225/add-on/src/lib/dnslink.js#L13-L14

**Comment by olizilla** [comment]
This is a good idea. I think the names could be used to inform people out the trade off.

- **Off**
- **Check after http request**
- **Check before http request**

If you reframe this as the user facing idea of "Load IPFS hosted sites over IPFS where possible" rather than "Enable DNSLink" then it's a no-brainer to enable it by default.

**Comment by lidel** [comment]
@olizilla  thank you, those are serious improvements. 

I feel now most of users won't need to "read more" to understand what is what:

screenshot_4
